lua/lib/lfospec.lua

```diff
@@ -0,0 +1,419 @@
+-- LFOs for general-purpose scripting
+-- @module lib.lfospec
+-- inspired by contributions from @markwheeler (changes), @justmat (hnds), and @sixolet (toolkit)
+-- added by @dndrks
+
+local LFO = {}
+LFO.__index = LFO
+
+local lfo_rates = {1/16,1/8,1/4,5/16,1/3,3/8,1/2,3/4,1,1.5,2,3,4,6,8,16,32,64,128,256,512,1024}
+local lfo_rates_as_strings = {"1/16","1/8","1/4","5/16","1/3","3/8","1/2","3/4","1","1.5","2","3","4","6","8","16","32","64","128","256","512","1024"}
+
+local update_freq = 96
+local params_per_entry = 14
+local lfos_all_loaded = false
+
+local rand_values;
+
+--- construct an LFO
+-- consumes one clock per LFO (norns has 100 clocks available for scripting)
+-- @tparam[opt] string shape The shape for this LFO (options: 'sine','saw','square','random'; default: 'sine')
+-- @tparam[opt] number min The minimum bound for this LFO (default: 0)
+-- @tparam[opt] number max The maximum bound for this LFO (default: 1)
+-- @tparam[opt] number depth The depth of modulation between min/max (range: 0.0 to 1.0; default: 0.0)
+-- @tparam[opt] string mode How to advance the LFO (options: 'clocked', 'free'; default: 'clocked')
+-- @tparam[opt] number period The timing of this LFO's advancement. If mode is 'clocked', argument is in beats. If mode is 'free', argument is in Hz.
```

Comment:
This might be tradition, but it feels a little strange to have a time-measure in one mode and a 1/time measure in the other. Maybe beats & seconds is better, since they're both time measures? I know for slow LFOs I always find myself taking the reciprocal in my head anyway if I'm trying to specify hz.

Reply:
def pulling from tradition here -- @tehn , what do you think?

Reply:
agreed! changed